<animateMotion> non-path animations should apply the 'rotate' attribute

Ahmad-S792 opened the pull request:
#### 50984dcaa2fdad2d3f8277b63cc473ca18aeee5c
<pre>
&lt;animateMotion&gt; non-path animations should apply the &apos;rotate&apos; attribute
<a href="https://bugs.webkit.org/show_bug.cgi?id=257907">https://bugs.webkit.org/show_bug.cgi?id=257907</a>
<a href="https://rdar.apple.com/problem/110915794">rdar://problem/110915794</a>

Reviewed by Nikolas Zimmermann.

Partial Merge: <a href="https://chromium.googlesource.com/chromium/src/+/7f43c0d43ff1b00697461a588118d69e8d86238a">https://chromium.googlesource.com/chromium/src/+/7f43c0d43ff1b00697461a588118d69e8d86238a</a>

&lt;animateMotion&gt; animations specified use the &apos;path&apos; attribute (or an
&lt;mpath&gt; etc) are not special in any way with respect to the &apos;rotate&apos;
attribute and its effects.

Refactor SVGAnimateMotionElement::calculateAnimatedValue() so that said
attribute is also applied in the non-path animation case as well.

In addition, this made `RotateMode` as enum class and used it throughout
code and removed `default` handling of RotateMode, since all cases are
handled properly.

* Source/WebCore/svg/SVGAnimateMotionElement.cpp:
(WebCore::SVGAnimateMotionElement::rotateMode const):
(WebCore::SVGAnimateMotionElement::calculateAnimatedValue):
* Source/WebCore/svg/SVGAnimateMotionElement.h:
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animateMotion-from-to-rotate-auto-expected.txt: Progression

Also added tests covering the degenerate &lt;animateMotion&gt; cases where the
from/to points share a coordinate (fromPoint.y == toPoint.y) or are
identical (zero delta), guarding against a NaN angle out of atan2.

* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animateMotion-from-to-rotate-auto-horizontal-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animateMotion-from-to-rotate-auto-horizontal.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animateMotion-from-to-rotate-auto-no-motion-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animateMotion-from-to-rotate-auto-no-motion.html: Added.

Canonical link: <a href="https://commits.webkit.org/314508@main">https://commits.webkit.org/314508@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/28682a158310bd1297aa8ced88c8de18281f5764

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/165639 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/39129 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/32710 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/174681 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/122894 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/d6248caf-bbaf-4b9d-9c90-ac8c9a0b4442) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/167505 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/39465 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/39133 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/128722 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/90645 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/50d30f75-1699-49bd-b4db-8198e221e76b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/168598 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/31054 "Found 2 new test failures: fast/events/touch/ios/touch-event-regions-layer-tree/mousemove-mouseup.html fast/forms/ios/drag-range-track.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/148814 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/109246 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/24999e78-8577-4e45-917b-eb41e4986d1d) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/30548 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-images/gradients-with-border.html imported/w3c/web-platform-tests/css/css-images/multiple-position-color-stop-linear.html (failure)") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/28726 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/22761 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/139984 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/26512 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/177205 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/30117 "Built successfully") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/28218 "Build is in progress. Recent messages:OS: Sequoia (15.7.3), Xcode: 26.2; Running apply-patch; Checked out pull request; Passed layout tests; 4 flakes 2 failures; Uploaded test results; 3 flakes") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/137045 "Passed tests") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/39198 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/32870 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/136978 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/37057 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/39886 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/148308 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/102378 "Built successfully") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/32265 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/25175 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/38397 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/111470 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/37793 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/3256 "Passed tests") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/38039 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/37937 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Ahmad-S792 committed:
<animateMotion> non-path animations should apply the 'rotate' attribute
https://bugs.webkit.org/show_bug.cgi?id=257907
rdar://problem/110915794

Reviewed by NOBODY (OOPS!).

Partial Merge: https://chromium.googlesource.com/chromium/src/+/7f43c0d43ff1b00697461a588118d69e8d86238a

<animateMotion> animations specified use the 'path' attribute (or an
<mpath> etc) are not special in any way with respect to the 'rotate'
attribute and its effects.

Refactor SVGAnimateMotionElement::calculateAnimatedValue() so that said
attribute is also applied in the non-path animation case as well.

In addition, this made `RotateMode` as enum class and used it throughout
code and removed `default` handling of RotateMode, since all cases are
handled properly.

* Source/WebCore/svg/SVGAnimateMotionElement.cpp:
(WebCore::SVGAnimateMotionElement::rotateMode const):
(WebCore::SVGAnimateMotionElement::calculateAnimatedValue):
* Source/WebCore/svg/SVGAnimateMotionElement.h:
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animateMotion-from-to-rotate-auto-expected.txt: Progression

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/83408fa82bba1d3c2e5ccc82d1b173bbd9d2d1e8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49821

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6d2eb07bb47ae8873e1611e6b51aeba7c94b05aa. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49821

karlcow commented on Source/WebCore/svg/SVGAnimateMotionElement.cpp:
```diff
@@ -223,26 +226,37 @@ void SVGAnimateMotionElement::calculateAnimatedValue(float percentage, unsigned
-    buildTransformForProgress(transform, percentage);
+        // Compute angle from fromPoint -> toPoint.
+        FloatPoint delta(m_toPoint.x() - m_fromPoint.x(), m_toPoint.y() - m_fromPoint.y());
+        angle = rad2deg(delta.slopeAngleRadians());
```
```
float FloatPoint::slopeAngleRadians() const
{
    return atan2f(m_y, m_x);
}
```

So I guess the returned value is always [-π ; +π] radians



There is similar code in 
https://github.com/WebKit/WebKit/blob/be6b53f67441fb461d9a861593bab437b07b8a51/Source/WebCore/rendering/svg/SVGMarkerData.h#L88-L113

and in 

https://github.com/WebKit/WebKit/blob/be6b53f67441fb461d9a861593bab437b07b8a51/Source/WebCore/svg/SVGPathParser.cpp#L440-L457

nikolaszimmermann replied:
Ok, as `atan2` is used, we need to make sure that neither x or y are null, otherwise we'll receive `NaN`, in many implementations. If we touch the code again, we should make sure we'll never get NaN values IMHO.

nikolaszimmermann replied:
Can you construct a testcase in which fromPoint x equals to toPoint x (or for y)?

nikolaszimmermann commented on Source/WebCore/svg/SVGAnimateMotionElement.cpp:
```diff
@@ -223,26 +226,37 @@ void SVGAnimateMotionElement::calculateAnimatedValue(float percentage, unsigned
+    case RotateAngle:
+        angle = 0;
+        break;
+    default:
```
```
    enum RotateMode {
        RotateAngle,
        RotateAuto,
        RotateAutoReverse
    };
```

Oh, indeed the switch already handles all cases, that means the default is useless. We should make the RotateMode an enum class as well -- maybe makes sense given that we touch this code now.

Ahmad-S792 replied:
Removed `default` handling and also updated to use enum class.

nikolaszimmermann reviewed (approved):
LGTM.

Ahmad-S792 committed:
<animateMotion> non-path animations should apply the 'rotate' attribute
https://bugs.webkit.org/show_bug.cgi?id=257907
rdar://problem/110915794

Reviewed by Nikolas Zimmermann.

Partial Merge: https://chromium.googlesource.com/chromium/src/+/7f43c0d43ff1b00697461a588118d69e8d86238a

<animateMotion> animations specified use the 'path' attribute (or an
<mpath> etc) are not special in any way with respect to the 'rotate'
attribute and its effects.

Refactor SVGAnimateMotionElement::calculateAnimatedValue() so that said
attribute is also applied in the non-path animation case as well.

In addition, this made `RotateMode` as enum class and used it throughout
code and removed `default` handling of RotateMode, since all cases are
handled properly.

* Source/WebCore/svg/SVGAnimateMotionElement.cpp:
(WebCore::SVGAnimateMotionElement::rotateMode const):
(WebCore::SVGAnimateMotionElement::calculateAnimatedValue):
* Source/WebCore/svg/SVGAnimateMotionElement.h:
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animateMotion-from-to-rotate-auto-expected.txt: Progression

Also added tests covering the degenerate <animateMotion> cases where the
from/to points share a coordinate (fromPoint.y == toPoint.y) or are
identical (zero delta), guarding against a NaN angle out of atan2.

* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animateMotion-from-to-rotate-auto-horizontal-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animateMotion-from-to-rotate-auto-horizontal.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animateMotion-from-to-rotate-auto-no-motion-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animateMotion-from-to-rotate-auto-no-motion.html: Added.

Canonical link: https://commits.webkit.org/314508@main
